allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py

```diff
@@ -29,76 +31,155 @@ class PytorchSeq2SeqWrapper(Seq2SeqEncoder):
-    def get_input_dim(self) -> int:
-        return self._module.input_size
+        self._stateful = stateful
+        self._max_batch_size = max_batch_size
+        self._states = None
 
-    def get_output_dim(self) -> int:
         try:
             is_bidirectional = self._module.bidirectional
         except AttributeError:
             is_bidirectional = False
-        return self._module.hidden_size * (2 if is_bidirectional else 1)
+        if is_bidirectional:
+            self._num_directions = 2
+        else:
+            self._num_directions = 1
+
+    def get_input_dim(self) -> int:
+        return self._module.input_size
+
+    def get_output_dim(self) -> int:
+        return self._module.hidden_size * self._num_directions
 
     def forward(self,  # pylint: disable=arguments-differ
                 inputs: torch.Tensor,
                 mask: torch.Tensor,
                 hidden_state: torch.Tensor = None) -> torch.Tensor:
 
+        if self._stateful and mask is None:
+            raise ValueError("Always pass a mask with stateful RNNs.")
+        if self._stateful and hidden_state is not None:
+            raise ValueError("Stateful RNNs provide their own initial hidden_state.")
+
         if mask is None:
             return self._module(inputs, hidden_state)[0]
 
         # In some circumstances you may have sequences of zero length.
-        # ``pack_padded_sequence`` requires all sequence lengths to be > 0, so here we
-        # adjust the ``mask`` so that every sequence has length at least 1. Then after
-        # running the RNN we zero out the corresponding rows in the result.
+        # ``pack_padded_sequence`` requires all sequence lengths to be > 0, so
+        # remove sequences of zero length before calling self._module, then fill with
+        # zeros.
 
         # First count how many sequences are empty.
         batch_size, total_sequence_length = mask.size()
         num_valid = torch.sum(mask[:, 0]).int().data[0]
 
-        # Force every sequence to be length at least one. Need to `.clone()` the mask
-        # to avoid a RuntimeError from shared storage.
-        if num_valid < batch_size:
-            mask = mask.clone()
-            mask[:, 0] = 1
+        if self._stateful and batch_size > self._max_batch_size:
+            raise ValueError("Got batch_size={0} but using a stateful RNN with max_batch_size={1}"
+                             "".format(batch_size, self._max_batch_size))
 
         sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
         sorted_inputs, sorted_sequence_lengths, restoration_indices = sort_batch_by_length(inputs,
                                                                                            sequence_lengths)
-        packed_sequence_input = pack_padded_sequence(sorted_inputs,
-                                                     sorted_sequence_lengths.data.tolist(),
+        packed_sequence_input = pack_padded_sequence(sorted_inputs[:num_valid, :, :],
```

Comment:
This is a good change and actually fixes a bug if you have sequences of length 1. Thanks for doing this..